File: Data Structures & Algorithms/three-integer-sum/submission-0.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();

        for(int i = 0; i< n-2; i++){
            for(int j = i+1; j<n-1; j++){
                for(int k = j+1; k < n; k++){

                    if(nums[i] + nums[j]+nums[k] == 0){
                        List<Integer> triplet = new ArrayList<>();
                        triplet.add(nums[i]);
                        triplet.add(nums[j]);
                        triplet.add(nums[k]);

                        if(!res.contains(triplet)){
                            res.add(triplet);
                        }
                    }
                }
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-0.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        def recursion(i, subset):
            if i == len(nums):
                res.append(subset.copy())
                return
            # include
            subset.append(nums[i])
            recursion(i + 1, subset)
            subset.pop()

            # exclude
            while i + 1 < len(nums) and nums[i] == nums[i + 1]:
                i += 1
            recursion(i + 1, subset)
            
        recursion(0, [])
        return res